### ISSUE TITLE
Instrument for PhET-iO

### ISSUE THREAD

**Issue by samreid** [phetsims/john-travoltage#182]
@emily-phet and @ariel-phet requested @jbphet to instrument John Travoltage for PhET-iO.

## Overview of instrumentation steps:
 - [x] [Understand the goal](#understand-the-goal)
 - [x] [Code review](#code-review)
 - [x] [Import Tandem to main.js](#import-tandem-to-main.js)
 - [x] [Create root tandem](#create-root-tandem)
 - [x] [Pass tandems to sim screens](#pass-tandems-to-sim-screens)
 - [x] [Pass tandems to model and view](#pass-tandems-to-model-and-view)
 - [x] [Pass tandems to all constructors](#pass-tandems-to-all-constructors)
   + See also [Testing while adding Tandem instances](#testing-while-adding-tandem-instances). Test with `?phetioValidateTandems=false` since common code components don't yet have tandems.
 - [x] [Pass tandem instances to common code components](#pass-tandem-instances-to-common-code-components)
   + See also [Testing while adding Tandem instances](#testing-while-adding-tandem-instances).
 - [x] [Replace types with tandemized equivalents](#replace-types-with-tandemized-equivalents)
 - [x] [Instrument any Uninstrumented Common Code Components Used by the Sim](#instrument-any-uninstrumented-common-code-components-used-by-the-sim)
 - [x] [Instrument new component types defined by the sim](#instrument-new-component-types-defined-by-the-sim)
 - [x] [Add simulation specific support](#add-simulation-specific-support)
 - [x] [Support dynamic state](#support-dynamic-state)
 - [x] Are random numbers using `phet.joist.random`, and all doing so after modules are declared (non-statically)?  For
 instance, the following methods (and perhaps others) should not be used: `Math.random`, `_.shuffle`, `_.sample`, `_.random`
 - [x] `undefined` values cannot be saved from the phet-io state save, sims should be written to use `null` instead
 - [x] Verify that the simulation works in all of the phet-io wrappers
 - [x] Set up a task to have the newly instrumented sim tested by the QA team
 - [x] Update these instructions if you find them to be incomplete, inconsistent or incorrect
 - [ ] Add your instrumented simulation to the spreadsheet here: https://docs.google.com/spreadsheets/d/1ToW8wc9zOFpMglIf63jB1j8YA9Qmh2dxNLH2Sm4qxEo/edit?usp=drive_web
- [x] instrument ElectronNode
- [x] instrument JohnTravoltageKeyboardHelpContent
- [x] instrument SparkNode
- [x] When setting state from instance proxies => launch, the particles don't update the same. (Will address this in https://github.com/phetsims/john-travoltage/issues/188 )

**Comment by samreid** [comment]
There was a bit of cruft leftover from prior work on instrumentation, I'll clean it up shortly and try to get it launching.

**Comment by samreid** [comment]
I've flagged 2 issues (see above) that should be closed as part of (or before starting) PhET-iO work.

I also committed enough fixes to get the simulation launching with `?phetioValidateTandems=false`.  I noticed Electron.js needs to switch to the 2-arg `addInstance` call.  Also, this sim will be a good opportunity for @jbphet to take the https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#support-dynamic-state "dynamic state" instructions for a test drive.

**Comment by samreid** [comment]
The majority of the remaining checkboxes indicate incomplete instrumentation for accessibility and sonification code, but it is unclear whether PhET-iO instrumentation of those features are necessary at this point.  It would be good to check in with @jbphet about sonification soon and for me to review the accessibility instrumentation.  Combined with addressing code review issues, we aren't too far from a QA test.  @jessegreenberg in #185 you said (regarding the code review issues):

> I will be working through many of these this week

Perhaps we should synchronize our timing so that your fixes can also get checked during QA and we can still have this ready soon for @emily-phet.

**Comment by samreid** [comment]
Regarding the code review checkbox, I addressed many code review issues, but left open issues that (a) do not pertain to the instrumentation or (b) which pertain to a11y or sonification or which I am not sure how to solve.

**Comment by samreid** [comment]
RC.1 is up here: http://www.colorado.edu/physics/phet/dev/html/john-travoltage/1.3.0-phetiorc.1/wrappers/index/

**Comment by samreid** [comment]
On hold until testing completes.

**Comment by samreid** [comment]
We published 1.3.0-phetio, but only for ETS.  In a few weeks after master stabilizes, we are going to roll out another version for our larger PhET-iO client base.  On hold for 3 weeks until @jessegreenberg is done working on changes in master.

**Comment by samreid** [comment]
@zepumph and @jessegreenberg what is the status of this issue?

**Comment by zepumph** [comment]
JT 1.4.0-phetio is published to the phet-io website, and just went through full QA testing for phet-io. I think we should close this and make a new issue to deliver it to our non ETS partners. @jessegreenberg are there any other thoughts you have?

**Comment by zepumph** [comment]
I'm closing this. The most current release is 1.4.2-phetio.

 https://phet-io.colorado.edu/sims/john-travoltage/1.4.2-phetio/wrappers/index/
